source/common/http/codes.cc

```diff
@@ -133,19 +154,63 @@ absl::string_view CodeStatsImpl::stripTrailingDot(absl::string_view str) {
   return str;
 }
 
-std::string CodeStatsImpl::join(const std::vector<absl::string_view>& v) {
-  if (v.empty()) {
-    return "";
+Stats::StatName CodeStatsImpl::upstreamRqGroup(Code response_code) const {
+  switch (enumToInt(response_code) / 100) {
+  case 1:
+    return upstream_rq_1xx_;
+  case 2:
+    return upstream_rq_2xx_;
+  case 3:
+    return upstream_rq_3xx_;
+  case 4:
+    return upstream_rq_4xx_;
+  case 5:
+    return upstream_rq_5xx_;
+  }
+  return upstream_rq_unknown_;
+}
+
+CodeStatsImpl::RequestCodeGroup::RequestCodeGroup(absl::string_view prefix,
+                                                  CodeStatsImpl& code_stats)
+    : code_stats_(code_stats), prefix_(std::string(prefix)),
+      stat_name_pool_(code_stats_.symbol_table_) {
+  for (uint32_t i = 0; i < NumHttpCodes; ++i) {
+    rc_stat_names_[i] = nullptr;
   }
-  auto iter = v.begin();
-  if (iter->empty()) {
-    ++iter; // Skip any initial empty prefix.
+
+  // Pre-allocate response codes 200, 404, and 503, as those seem quite likely.
+  // We don't pre-allocate all the HTTP codes because the first 127 allocations
+  // are likely to be encoded in one byte, and we would rather spend those on
+  // common components of stat-names that appear frequently.
+  statName(Code::OK);
+  statName(Code::NotFound);
+  statName(Code::ServiceUnavailable);
+}
+
+Stats::StatName CodeStatsImpl::RequestCodeGroup::statName(Code response_code) {
+  // Take a lock only if we've never seen this response-code before.
+  uint32_t rc_int = static_cast<uint32_t>(response_code);
+  RELEASE_ASSERT(rc_int < NumHttpCodes, absl::StrCat("Unexpected http code: ", rc_int));
```

Comment:
It seems like above we are handling "unknown." If so, it seems like we shouldn't crash here. Can we be consistent and probably not crash given this depends on external input?

Reply:
Agreed. Great catch, thanks.